File: rust/herb-linter/src/rules/html_no_aria_hidden_on_focusable.rs
use std::collections::HashSet;
use std::sync::LazyLock;

use crate::utils::tag_utils::{get_attribute, get_static_attribute_value, get_tag_name_from_open_tag, has_aria_hidden_true, has_attribute, tag_name_location};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(NoAriaHiddenOnFocusableVisitor);
define_parser_rule!(
  HTMLNoAriaHiddenOnFocusableRule,
  "html-no-aria-hidden-on-focusable",
  Error,
  NoAriaHiddenOnFocusableVisitor
);

static INTERACTIVE_ELEMENTS: LazyLock<HashSet<&'static str>> =
  LazyLock::new(|| ["button", "summary", "input", "select", "textarea", "a"].into_iter().collect());

fn get_tab_index_value(node: &HTMLOpenTagNode) -> Option<i32> {
  get_attribute(node, "tabindex")
    .and_then(|attribute| get_static_attribute_value(attribute))
    .and_then(|value| value.parse::<i32>().ok())
}

fn is_focusable(node: &HTMLOpenTagNode) -> bool {
  let tag_name = match get_tag_name_from_open_tag(node) {
    Some(name) => name,
    None => return false,
  };

  let tag_lower = tag_name.to_lowercase();
  let tab_index = get_tab_index_value(node);

  if tag_lower == "a" {
    let has_href = has_attribute(node, "href");

    if !has_href {
      return tab_index.map(|v| v >= 0).unwrap_or(false);
    }

    return tab_index.map(|v| v >= 0).unwrap_or(true);
  }

  if INTERACTIVE_ELEMENTS.contains(tag_lower.as_str()) {
    tab_index.map(|v| v >= 0).unwrap_or(true)
  } else {
    tab_index.map(|v| v >= 0).unwrap_or(false)
  }
}

impl Visitor for NoAriaHiddenOnFocusableVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    if has_aria_hidden_true(node) && is_focusable(node) {
      self.add_offense(
        "Elements that are focusable should not have `aria-hidden=\"true\"` because it will cause confusion for assistive technology users.".to_string(),
        tag_name_location(node),
      );
    }

    self.walk_html_open_tag_node(node);
  }
}

File: javascript/packages/linter/src/rules/html-no-aria-hidden-on-focusable.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor, isKeyboardFocusableElement } from "./rule-utils.js"
import { getAttributeValue, findAttributeByName, getAttributes } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLOpenTagNode, ParseResult } from "@herb-tools/core"

class NoAriaHiddenOnFocusableVisitor extends BaseRuleVisitor {
  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    this.checkAriaHiddenOnFocusable(node)
    super.visitHTMLOpenTagNode(node)
  }

  private checkAriaHiddenOnFocusable(node: HTMLOpenTagNode): void {
    if (!this.hasAriaHiddenTrue(node)) return

    if (isKeyboardFocusableElement(node)) {
      this.addOffense(
        `Elements that are focusable should not have \`aria-hidden="true"\` because it will cause confusion for assistive technology users.`,
        node.tag_name!.location,
      )
    }
  }

  private hasAriaHiddenTrue(node: HTMLOpenTagNode): boolean {
    const attributes = getAttributes(node)
    const ariaHiddenAttr = findAttributeByName(attributes, "aria-hidden")

    if (!ariaHiddenAttr) return false

    const value = getAttributeValue(ariaHiddenAttr)

    return value === "true"
  }

}

export class HTMLNoAriaHiddenOnFocusableRule extends ParserRule {
  static ruleName = "html-no-aria-hidden-on-focusable"
  static introducedIn = this.version("0.6.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoAriaHiddenOnFocusableVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
